internal/app/machined/pkg/controllers/config/acquire.go

```diff
@@ -0,0 +1,437 @@
+					},
+				)
+
+				return err
+			}
+
+			if cfg != nil {
+				// apply config
+				if err = ctrl.ConfigSetter.SetConfig(cfg); err != nil {
+					return fmt.Errorf("failed to set config: %w", err)
+				}
+			}
+
+			if newState == nil {
+				// wait for reconcile event, keep running in the same state
+				break
+			}
+
+			currentState = newState
+		}
+
+		r.ResetRestartBackoff()
+	}
+}
+
+// stateDisk acquires machine configuration from disk (STATE partition).
+//
+// Transitions:
+//
+//	--> platform: no config found on disk, proceed to platform
+//	--> maintenanceEnter: config found on disk, but it's incomplete, proceed to maintenance
+//	--> done: config found on disk, and it's complete
+func (ctrl *AcquireController) stateDisk(ctx context.Context, r controller.Runtime, logger *zap.Logger) (stateMachineFunc, config.Provider, error) {
+	cfg, err := ctrl.loadFromDisk(logger)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	switch {
+	case cfg == nil:
+		// no config loaded, proceed to platform
+		return ctrl.statePlatform, nil, nil
+	case cfg.CompleteForBoot():
+		// complete config, we are done
+		return ctrl.stateDone, cfg, nil
+	default:
+		// incomplete config, proceed to maintenance
+		return ctrl.stateMaintenanceEnter, cfg, nil
+	}
+}
+
+// loadFromDisk is a helper function for stateDisk.
+func (ctrl *AcquireController) loadFromDisk(logger *zap.Logger) (config.Provider, error) {
+	logger.Debug("loading config from STATE", zap.String("path", ctrl.ConfigPath))
+
+	_, err := os.Stat(ctrl.ConfigPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// no saved machine config
+			return nil, nil
+		}
+
+		return nil, fmt.Errorf("failed to stat %s: %w", ctrl.ConfigPath, err)
+	}
+
+	cfg, err := configloader.NewFromFile(ctrl.ConfigPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load config from STATE: %w", err)
+	}
+
+	// this is legacy flow with persistence, it should be deprecated & removed
+	if !cfg.Persist() {
+		logger.Info("config persistence is disabled, ignoring stored machine config")
+
+		return nil, nil
+	}
+
+	warnings, err := cfg.Validate(ctrl.ValidationMode)
+	if err != nil {
+		return nil, fmt.Errorf("failed to validate on-disk config: %w", err)
+	}
+
+	for _, warning := range warnings {
+		logger.Warn("config validation warning", zap.String("warning", warning))
+	}
+
+	return cfg, nil
+}
+
+// statePlatform acquires machine configuration from the platform source.
+//
+// Transitions:
+//
+//	--> maintenanceEnter: config loaded from platform, but it's incomplete, or no config from platform: proceed to maintenance
+//	--> done: config loaded from platform, and it's complete
+func (ctrl *AcquireController) statePlatform(ctx context.Context, r controller.Runtime, logger *zap.Logger) (stateMachineFunc, config.Provider, error) {
+	cfg, err := ctrl.loadFromPlatform(ctx, logger)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	switch {
+	case cfg == nil:
+		fallthrough
+	case !cfg.CompleteForBoot():
```

Comment:
so any state that's not CompleteForBoot is assumed to be done I guess, wonder if there's some state we'd need to handle :thinking: